src/journal/sd-journal.c

```diff
@@ -2661,13 +2661,28 @@ _public_ int sd_journal_wait(sd_journal *j, uint64_t timeout_usec) {
         assert_return(!journal_pid_changed(j), -ECHILD);
 
         if (j->inotify_fd < 0) {
+                Iterator i;
+                JournalFile *f;
 
                 /* This is the first invocation, hence create the
                  * inotify watch */
                 r = sd_journal_get_fd(j);
                 if (r < 0)
                         return r;
 
+                /* Server might have done some vacuuming while we weren't watching.
+                   Get rid of the deleted files now so they don't stay around indefinitely. */
+                ORDERED_HASHMAP_FOREACH(f, j->files, i) {
+                        r = journal_file_fstat(f);
+                        if (r < 0) {
+                                log_debug_errno(r,"Failed to fstat() journal file '%s' : %m", f->path);
+                                continue;
+                        }
+
+                        if (f->last_stat.st_nlink <= 0)
```

Comment:
nit: nlink_t is unsigned, so it can't be less than 0.

Reply:
We have such check in one more place. I will send followup PR to fix it in both places.

Reply:
Also strictly speaking POSIX doesn't specify singleness for nlink_t.

```
nlink_t, uid_t, gid_t, and id_t shall be integer types.
```

Reply:
systemd runs on Linux :-)

Reply:
@poettering likes to write checks this way. The justification is that if the type is changed to signed later on, the check will still DTRT. Personally, I'd prefer == 0. But since both versions are equivalent, we should treat it as a matter of style, up to the author.

Reply:
`<=` is already used in one more place so let's just keep it then (considering that @poettering also has slight preference for this style).